File: trunk/Noxa.Emulation.Psp.Bios.FastHLE/Modules/sceDmac.cpp
// ----------------------------------------------------------------------------
// PSP Player Emulation Suite
// Copyright (C) 2006 Ben Vanik (noxa)
// Licensed under the LGPL - see License.txt in the project root for details
// ----------------------------------------------------------------------------

#include "Stdafx.h"
#include "sceDmac.h"
#include "Kernel.h"
#include <string.h>
#include <assert.h>

using namespace System;
using namespace System::Diagnostics;
using namespace Noxa::Emulation::Psp;
using namespace Noxa::Emulation::Psp::Bios;
using namespace Noxa::Emulation::Psp::Bios::Modules;

int sceDmacMemcpyN( byte* memory, int dest, int source, int size );

void* sceDmac::QueryNativePointer( uint nid )
{
	switch( nid )
	{
	case 0x617F3FE6:
		return &sceDmacMemcpyN;
	};

	return 0;
}

#pragma unmanaged
int sceDmacMemcpyN( byte* memory, int dest, int source, int size )
{
	// Only support main memory copies - anything else can DIAF
	assert( ( dest >= MainMemoryBase ) && ( dest < MainMemoryBound ) );
	assert( ( source >= MainMemoryBase ) && ( source < MainMemoryBound ) );

	memcpy(
		( memory + ( dest - MainMemoryBase ) ),
		( memory + ( source - MainMemoryBase ) ),
		size );

	return dest;
}
#pragma managed

int sceDmac::sceDmacMemcpy( IMemory^ memory, int dest, int source, int size )
{
	byte* ptr = ( byte* )memory->MainMemoryPointer;

	memcpy(
		( ptr + ( dest - MainMemoryBase ) ),
		( ptr + ( source - MainMemoryBase ) ),
		size );

	return dest;
}

File: Noxa.Emulation.Psp.Bios.ManagedHLE/Modules/sceDmac.cs
// ----------------------------------------------------------------------------
// PSP Player Emulation Suite
// Copyright (C) 2006 Ben Vanik (noxa)
// Licensed under the LGPL - see License.txt in the project root for details
// ----------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Text;

using Noxa.Utilities;
using Noxa.Emulation.Psp;
using Noxa.Emulation.Psp.Bios;
using Noxa.Emulation.Psp.Cpu;

namespace Noxa.Emulation.Psp.Bios.ManagedHLE.Modules
{
	class sceDmac : Module
	{
		#region Properties

		public override string Name
		{
			get
			{
				return "sceDmac";
			}
		}

		#endregion

		#region State Management

		public sceDmac( Kernel kernel )
			: base( kernel )
		{
		}

		public override void Start()
		{
		}

		public override void Stop()
		{
		}

		#endregion

		[Stateless]
		[BiosFunction( 0x617F3FE6, "sceDmacMemcpy" )]
		// manual add
		public int sceDmacMemcpy( int dest, int source, int size )
		{
			return Module.NotImplementedReturn;
		}

	}
}
